### PR TITLE
Remove `WebUtility` use and handle `IHtmlContent` return values from view components

### PR DESCRIPTION
**Pull request by dougbu**
- #3571 and part of #3123
- split `ContentViewComponentResult` in two
- add `HtmlEncoder` to `ViewComponentContext`
- remove use of `WebUtility.HtmlEncode()` and `HtmlDecode()`

nits: remove unused `using`s in files I had open

### CODE REVIEW

**Comment by dougbu** [comment]
Performance wasn't the goal of this change but avoiding `string` allocations when HTML encoding was not necessary did have some benefit. Numbers below are for 3000 requests and a simple site with one view component that returns a short `string`.

Before:
before

After:
after

**Comment by rynowak** [comment]
@dougbu - have you puzzled out there the reduction in allocations came from? The total number indicates about .7mb was shaved off, the difference in byte[] was about .5mb. 

byte[] is generally the 'experimental error' - it's variable based on the amount of server load, and the amount of output written (until kestrel finishes the SocketOutput work).

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponent.cs [hunk 1]
```diff
@@ -201,15 +201,15 @@ public ICompositeViewEngine ViewEngine
         /// Returns a result which will render HTML encoded text.
         /// </summary>
         /// <param name="content">The content, will be HTML encoded before output.</param>
-        /// <returns>A <see cref="ContentViewComponentResult"/>.</returns>
-        public ContentViewComponentResult Content(string content)
+        /// <returns>A <see cref="StringViewComponentResult"/>.</returns>
+        public StringViewComponentResult Content(string content)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponent.cs]: Why? This is meant to be an analouge of `ContentResult`

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs [hunk 2]
```diff
@@ -229,7 +238,7 @@ private ViewComponentDescriptor SelectComponent(Type componentType)
                 throw new ArgumentNullException(nameof(descriptor));
             }
 
-            var context = new ViewComponentContext(descriptor, arguments, _viewContext, writer);
+            var context = new ViewComponentContext(descriptor, arguments, _htmlEncoder, _viewContext, writer);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs]: Should we make the encoder part of the `ViewContext` instead? It seems to show up everywhere that's used.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponent.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponent.cs]: I like the cleanup of this class, but not the rename.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs]: Let's consider this later, once other aspects of HTML encoding settle.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentHelper.cs]: Agreed, I just want to pose the question.

**Comment by dougbu** [comment]
> have you puzzled out there the reduction in allocations came from?

Nah. `WebUtility` looks suspect but I'm not going to dive deeper.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponent.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponent.cs]: Restored old name for the `string` result.

**Comment by dougbu** [comment]
:up::date:

**Comment by rynowak** [comment]
:shipit:

**Comment by dougbu** [comment]
eb70b1c

**Pull request closed without merging**